### PR TITLE
Fixes issue #7543

### PR DESCRIPTION
**Pull request by hbina**
Added a helper function for `util::hex::hash_64` that uses streams
the content instead of reading through the entire content in one
go.

Edit: Should I add test cases for this?

Edit2: Per #7543

### CODE REVIEW

File: src/cargo/util/hex.rs [hunk 1]
```diff
@@ -21,6 +23,19 @@ pub fn hash_u64<H: Hash>(hashable: H) -> u64 {
...
 
+pub fn hash_u64_file(mut file: &File) -> u64 {
+    let mut hasher = SipHasher::new_with_keys(0, 0);
+    let mut buf = [0; 64 * 1024];
+    loop {
+        let n = file.read(&mut buf).unwrap();
+        if n == 0 {
+            break;
+        }
+        hasher.write(&buf);
```
Comment (Eh2406) [inline src/cargo/util/hex.rs]: comparing it to https://github.com/rust-lang/cargo/blob/ab2b2c0b05a5ec8f15b17433a56b3a8b70f4cc2a/src/cargo/util/sha256.rs#L20-L29
I'd have expected `hasher.write(&buf[..n]);`

**Comment by Eh2406** [comment]
The red CI is unrelated (cc 54242f50f8582c2d891f8552c561e244c000c5d8d2a4d8bfb6696114a784708d if I feal like investigating.)

Over all this look good with one nit. But I don't know the context well. So,
@bors r? @ehuss

**Comment by hbina** [comment]
I am getting this error
```
hbina@akarin:~/Documents/test_cargo$ cargo publish --dry-run
    Updating crates.io index
warning: manifest has no description, license, license-file, documentation, homepage or repository.
See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata for more info.
   Packaging test_cargo_7543 v0.1.0 (/home/hbina/Documents/test_cargo)
   Verifying test_cargo_7543 v0.1.0 (/home/hbina/Documents/test_cargo)
   Compiling test_cargo_7543 v0.1.0 (/home/hbina/Documents/test_cargo/target/package/test_cargo_7543-0.1.0)
    Finished dev [unoptimized + debuginfo] target(s) in 1.56s
error: failed to verify package tarball

Caused by:
  Source directory was modified by build.rs during cargo publish. Build scripts should not modify anything outside of OUT_DIR.
Added: /home/hbina/Documents/test_cargo/target/package/test_cargo_7543-0.1.0/tests/generated/large.txt

To proceed despite this, pass the `--no-verify` flag.
```

So I am not sure if it actually works.

File: src/cargo/util/hex.rs [hunk 2]
```diff
@@ -21,6 +23,19 @@ pub fn hash_u64<H: Hash>(hashable: H) -> u64 {
     hasher.finish()
 }
 
+pub fn hash_u64_file(mut file: &File) -> u64 {
+    let mut hasher = SipHasher::new_with_keys(0, 0);
+    let mut buf = [0; 64 * 1024];
+    loop {
+        let n = file.read(&mut buf).unwrap();
```
Comment (FauxFaux) [inline src/cargo/util/hex.rs]: The original code returns errors, instead of panicing.

File: src/cargo/util/hex.rs (hunk 1 above)
Comment (hbina) [reply src/cargo/util/hex.rs]: Oof, sorry about that.

File: src/cargo/util/hex.rs (hunk 2 above)
Comment (hbina) [reply src/cargo/util/hex.rs]: Yeah, I have changed it but now all the tests are failing..

**Comment by hbina** [comment]
Thanks for reviewing!

**Pull request merged**